tuna/fin_class.py

```diff
@@ -291,62 +317,68 @@ def dump_json(self, outfile, to_file=True):
 
     return True
 
-  def run(self, outfile=None, to_file=True):
+  def prep_fin_input(self, outfile=None, to_file=True):
     """Main function in Fin that produces Fin input file"""
 
     self.cnx = self.machine.connect(self.chk_abort_file())
-    ret = ''
+    ret = False
     if outfile is None:
       outfile = "fin_input.json"
     if self.create_dumplist():
       ret = self.dump_json(outfile, to_file)
     else:
-      self.logger.error("Could not create dumplist for Fin input file")
+      self.logger.warning("Could not create dumplist for Fin input file")
 
     return ret
 
+  def insert_applicability(self, session, json_in):
+    """write applicability to sql"""
+    _, solver_id_map_h = get_solver_ids()
+    for elem in json_in:
+      if "applicable_solvers" in elem.keys():
+        #remove old applicability
+        query = 'delete from {} where config={} and session={}'.format(
+            self.dbt.solver_app.__tablename__, elem["input"]["config_tuna_id"],
+            self.session_id)
+        session.execute(query)
+        if not elem["applicable_solvers"]:
+          self.logger.warning("No applicable solvers for %s",
+                              elem["input"]["config_tuna_id"])
+        for solver in elem["applicable_solvers"]:
+          try:
+            new_entry = self.dbt.solver_app(
+                solver=solver_id_map_h[solver],
+                config=elem["input"]["config_tuna_id"],
+                session=self.session_id)
+            session.add(new_entry)
+          except KeyError:
+            self.logger.warning('Solver %s not found in solver table', solver)
+            self.logger.info("Please run 'go_fish.py --update_solver' first")
+            return False
+
+    self.logger.info('Commit bulk transaction, please wait')
+    session.commit()
```

Comment:
this needs to be in a try catch, possibly rollback if we are retrying

Reply:
It is, take a look at the session_retry function